jconstraints-core/src/main/java/gov/nasa/jpf/constraints/smtlibUtility/smtsolver/SMTCMDContext.java

```diff
@@ -140,24 +147,49 @@ public Result solve(Valuation result) {
         } else if (l.equals("unsat")) {
           res = Result.UNSAT;
           if (unsatCoreSolver) {
+            unsatCoreLastRun = new ArrayList<>();
+            unsatCoreLastRun.add(currentRun);
             ctx.getUnsatCore();
-            String core = "";
-            while (!solverOutput.ready() && p.isAlive()) {
-              Thread.sleep(10);
-            }
-            while (solverOutput.ready() && (line = solverOutput.readLine()) != null) {
-              core += line;
-              core += "\n";
-            }
-            analyzeCore(core);
+            analyzeCore(readProcessOutput(ctx, solverOutput));
           }
         }
       }
-      return res;
-    } catch (IOException | InterruptedException e) {
-      e.printStackTrace();
     }
-    return Result.ERROR;
+    return res;
+  }
+
+  protected static String readProcessOutput(
+      SMTLibExportGenContext context, BufferedReader processOut)
+      throws ExecutionException, InterruptedException, TimeoutException {
+    StringBuilder output = new StringBuilder();
+    String line;
+    String uid = "a" + UUID.randomUUID();
+    boolean matchedUid = false;
+    ExecutorService threadExecutor = Executors.newCachedThreadPool();
+    SimpleTimeLimiter timeLimiter = SimpleTimeLimiter.create(threadExecutor);
+    Callable<String> readLineCallable = processOut::readLine;
+
+    context.echo(uid);
+    try {
+      while (!matchedUid
+          && (line = timeLimiter.callWithTimeout(readLineCallable, 500L, TimeUnit.MILLISECONDS))
+              != null) {
+        if (line.equals(uid)) {
+          matchedUid = true;
+        }
+        output.append(line);
+        output.append("\n");
+      }
+      if (output.toString().contains("Parse Error")) {
+        throw new IllegalStateException("Parse Error");
+      }
+      if (!matchedUid) {
+        throw new IllegalStateException("Incomplete Output (Did not match UID)");
+      }
+      return output.toString().replace(uid, "").trim();
```

Comment:
The semantics here are odd. Only whitespace at the beginning and end of output are trimmed, and during one execution, a newline is immediately re-added.